### DISCUSSION TITLE
Puppeteer Problems with GC Functions

### DISCUSSION THREAD

**Discussion by NightLiveAppDev** [AntlerVC/firetable#1312] [Support / Q&A] [answered]
Hello everyone,

I'm currently developing a project using Rowy. This project involves dynamically adjusting SVGs based on Firestore data, then converting these SVGs to PNGs using Puppeteer. The resulting PNGs should then be stored in Firebase Cloud Storage.

The function is running on Node.js 18 and the Puppeteer version used is 15.5.0. I've also allocated 2GB of memory and set the timeout to 540 seconds.

Originally, I intended to implement this functionality with Google Cloud Functions and leverage the Rowy Task deployment mechanism. However, I've been encountering several challenges.[[Maybe related](https://issuetracker.google.com/issues/266279679?pli=1)] Specifically, I had difficulties establishing a connection to Chrome in the Cloud Function environment, despite using puppeteer.launch(). The error I encountered was: "TimeoutError: Timed out after 30000 ms while trying to connect to the browser! Only Chrome at revision r1022525 is guaranteed to work." (I'm not sure, if [here](https://github.com/puppeteer/puppeteer/issues/4796) is some hint to solve this..)

Given these challenges, I am now considering transitioning the functionality to Google Cloud Run, as I've discovered [here](https://github.com/puppeteer/puppeteer/blob/main/docs/troubleshooting.md#running-puppeteer-on-google-cloud-functions) that it might be more suited for running applications like Puppeteer. I am hoping to find an effective way to migrate this functionality, considering that Rowy currently does not directly support Google Cloud Run deployment.

Here is a snippet of the Rowy Task I've been working with and where the error gets thrown:

```
const extensionBody: TaskBody = async ({ row, db, change, ref, logging, storage }) => {
...
const puppeteer = require("puppeteer@15.5.0");

async function convertSvgToPng(svg) {
  const browser = await puppeteer.launch(); // The cause of the TimeoutError
  const page = await browser.newPage();
  await page.setContent(svg);
  const element = await page.$('svg');
  const png = await element.screenshot({ encoding: 'binary' });
  await browser.close();
  return png;
};
...
const image = await convertSvgToPng(Buffer.from(editedSvgString).toString());
...
```

In this forum post, I wanted to discuss my experiences and pose a few questions to the community:

- Has anyone successfully implemented Puppeteer in a Google Cloud Function, especially within the context of Rowy Tasks?
- Has anyone used Google Cloud Run for such purposes and can share their experiences compared to Google Cloud Functions?
- Could Rowy potentially extend its functionality to support Google Cloud Run deployments?
- Are there other services or technologies I should consider to implement this functionality?

I look forward to your answer :)

**Comment by BeeBombshell** [comment, marked as answer]
Hey @NightLiveAppDev, with our recent release, we upgraded the node version to 18, which has made Puppeteer incompatible. Unfortunately, at the moment, there's nothing we can do about this. However, you could try using another Node library that is compatible with the current version. Additionally, we're actively working on some features that could potentially make Puppeteer integration possible in the future.
